Obj-C updates

danomatika opened the pull request:
This PR includes some (badly needed) updates to the Obj-C wrapper:

* updated min deployment target to iOS 9
* improved sample rate handling, pd sample rate is now independent of hardware sample rate
* added IO buffering when audio unit sample rate conversion is being performed (#177)
* added new configure methods to specify separate number of input and output channels
* added easy to use audio session category options properties, ie. Bluetooth support, AirPlay, etc (helps with #249), AirPlay not tested yet
* added convenience support for configuring Record and MultiRoute categories
* basic AudioUnit v2 to v3 bridge (#259)\*
* PdBase message queuing can now be overridden by new `initializeWithQueue:` class method
* formatting improvements and documentation porting from z_libpd.h
* objc iOS pdtest now displays an AVRoutePicker button on iOS 11+

\* _This is **not** "open PdAudioUnit in Garageband" support as that requires more extensive multi instance handling work within libpd/pure-data sources. It *should*, however, allow for dropping a single PdAudioUnit instance into an AUGraph inside an application._

Only _possibly_ breaking changes are:

* PdAudioController readonly "numberChannels" property now "inputChannels" and "outputChannels"
* PdAudioController readonly "mixingEnabled" property is now readwrite "mixWithOthers"
* PdAudioUnit init: no longer allowed due to AUAudioUnitV2Bridge, use initWithComponentDescription:options:error: instead
* PdAudioUnit "_blockSizeAsLog" ivar renamed to "_blockFramesAsLog"

Additional work ideas (may not get to these):

* [ ] ~~PdAudioUnit macOS AUHAL support~~
* [ ] ~~Swift support and test~~
* [ ] ~~macOS - iOS catalyst support~~
EDIT: Did *not* get to any of these.

Tested so far on:
* iPhone 5S (iOS 12)
* iPhone 6S (iOS 13)

With:
* headphones
* a mono Mifa A1 Bluetooth speaker
* stereo Mpow S11 Bluetooth headphones(+ mono mic)
* full duplex stereo Roland Edirol UA-25 (trusty old USB 1.1. device)

danomatika commented:
Bluetooth input *and* output does work with the PlayAndRecord category while setting allowsBluetooth and allowsBluetoothA2DP, but only when the sample rate is a multiple of 8k, ala 48k. Setting 44.1k brings all sorts of buffer issues, which I am looking into now.

danomatika commented:
I notice now in the Apple docs, that it is recommended to *always* use the hardware sample rate with the RemoteIO audio unit. Not doing this *seems* to work. What doesn't work is specifying different numbers of input and output channels.

There may need to be a further fix such as rebuilding PdAudioUnit as a separate, reusable audio unit plugged into RemoteIO input and outputs separately.

danomatika commented:
After reading the docs some more, setting the internal input and output sample rate for the RemoteIO audio unit should be fine as this configures a set of built-in format converters.

As for the Bluetooth buffering issues, I tracked it down to the input frames being different from the output frames in the render callback which was not being handled.

danomatika commented:
With the latest commits, the sample rate conversion buffering and Bluetooth are now working well in my testing.

I also added:
* ability to override automatic audio session activation with a property as it can be slow on app start with Bluetooth, in which case it would be better to configure then activate in a different thread
* pass -1 for channel value to have channels picked up automatically, including between IO route changes
* prefers stereo to mono by default so mono speakers and Bluetooth output will get stereo mix down, this can be overridden by a property

Pinging for @cerupcat @satoshimuraki @AliMunchkin @fomalsd @virusys @dizzybanjo testing...

It would *great* if someone could test with a multi-channel USB 2 device. My older stereo USB 1 device works fine, but I'm hoping this work now enables *easy* large IO support as it has been a feature request for [PdParty](http://danomatika.com/code/pdparty).

danomatika commented:
Opps, testing latest commits on 6S again and now seeing buffer size issues. Here we go again.

danomatika commented:
Ok, there was a math error that should be fixed now.

virusys commented:
Exciting to see all these changes :) 

I have not been able to do an exhaustive test yet with all of my available devices/speakers but I did run `pdtest` on all of my iOS devices this morning just using their internal speakers. Here's the list of what I currently have available, if you would like me to try anything specific (other than what you have already mentioned re: USB 2.0 devices) please let me know!

- iPhone 6S, iOS 13.0
- iPad Pro (12.9-inch), iOS 13.3.1
- iPad (7th Generation), iOS 13.3
- iPad Air, iOS 12.4.4
- iPad (4th Generation), iOS 10.3.3

And external audio devices:

- Mpow Judge (unfortunately in bad shape, one of the ears is broken but the unit can still connect)
- Steinberg UR22mkII (connect with Camera Connection Kit)
- VTOP USB 2.0 Audio Capture Card

danomatika commented:
The iOS pdtest should have Bluetooth enabled. You can also try multi-channel capability by configuring input and output channels to -1 in ViewController.m. It should then use whatever input and output channels are available and automatically switch when the audio route is changed. pdtest also has a small button in the bottom left corner to bring up the iOS 11+ route picker popover.

> On Jan 21, 2020, at 5:31 PM, virusys <notifications@github.com> wrote:
> 
> Exciting to see all these changes :)
> 
> I have not been able to do an exhaustive test yet with all of my available devices/speakers but I did run pdtest on all of my iOS devices this morning just using their internal speakers. Here's the list of what I currently have available, if you would like me to try anything specific (other than what you have already mentioned re: USB 2.0 devices) please let me know!
> 
> iPhone 6S, iOS 13.0
> iPad Pro (12.9-inch), iOS 13.3.1
> iPad (7th Generation), iOS 13.3
> iPad Air, iOS 12.4.4
> iPad (4th Generation), iOS 10.3.3
> And external audio devices:
> 
> Mpow Judge (unfortunately in bad shape, one of the ears is broken but the unit can still connect)
> Steinberg UR22mkII (connect with Camera Connection Kit)
> VTOP USB 2.0 Audio Capture Card
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/libpd/libpd/pull/283?email_source=notifications&email_token=AADVK7LYEPFINXUAQOAHVMLQ64PNHA5CNFSM4KHYIVW2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJQLVPI#issuecomment-576764605>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AADVK7L7D4CAZJFQP7ZFN4LQ64PNHANCNFSM4KHYIVWQ>.
> 

--------
Dan Wilcox
@danomatika <http://twitter.com/danomatika>
danomatika.com <http://danomatika.com/>
robotcowboy.com <http://robotcowboy.com/>

danomatika commented:
Latest commits improve sample rate conversion detection and documentation. The PdAudioController audio session notification handlers are documented so they can be overridden in subclasses for custom handling.

danomatika commented:
ping @monkeyswarm

monkeyswarm commented:
Thanks for all this, @danomatika ! 
I have some devices with older iOS versions on them, I'll check it out.

danomatika commented:
Has anyone had a chance to test this yet? I'd like to get some confirmation so I can put out a new release.

virusys commented:
Hey sorry for the delay, I’ve been pretty under the weather the past week but I do actually need to take a look at these updates (specifically the bluetooth related stuff) for a contract I’m working on; hoping to get it out of the way in the next couple of days if my health permits.

monkeyswarm commented:
Plan to look at it this week. Note that my primary/newest device is an iPhone 6s (I’ve been clinging to that headphone jack).

Sent from my iPhone

> On Jan 28, 2020, at 2:34 PM, virusys <notifications@github.com> wrote:
> 
> ﻿
> Hey sorry for the delay, I’ve been pretty under the weather the past week but I do actually need to take a look at these updates (specifically the bluetooth related stuff) for a contract I’m working on; hoping to get it out of the way in the next couple of days if my health permits.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

danomatika commented:
Sample rate & buffering overview in PdAudioUnit:
* Uses it's own internal sample rate, independent of the session's
* Sets up sample rate conversion built into RemoteIO's input and output
* Tries to detect if either the input or output sample rate is *not* a multiple of its internal sample rate (or vice versa). If not, as in the case of 44.1k : 48k, it allocates input and output ring buffers.
* When buffering, it will pad the input buffer with empty pd sample blocks if the number of input sample frames is less than the required number of output sample frames, aka buffer size jitter This should only happen once or twice after the audio unit is started and, in my testing, is really only adding a 1 block (64 sample) delay to the input stream. _If you pad only the *exact* number of samples needed, you run the risk of always padding every other frame which results in a subtle zipper noise._
 * When not buffering, simple processes samples in and out as before as input and output buffer sizes should match and are multiples of Pd's block size.
* PdAudioUnit listens to notifications if it's sample rate or number of input or output channels is being changed, aka by an AUGraph
* PdAudioController listens for audio route changes (plug / unplug, etc) and reconfigures the internal PdAudioUnit, this automatically handles connecting Bluetooth, etc

The result should be click & zipper-free audio, both input and output, while Pd maintains it's internal sample rate. Running `samples/objc/ios/pdtest` should result in the same sine wave output frequency no matter the combinations of sample rates. I also added a button to pdtest which opens the audio route picker popover so you can switch between Bluetooth and internal audio. Airplay should work, but I've not tested it directly yet.

To confirm if buffering is happening (or not), uncomment `AU_DEBUG_VERBOSE` at the top of `libpd/objc/AudioHelpers.h" and look for a "inited sample rate conversion buffers" line after the audio unit is (re)configured. If the line is not there, then buffering is not needed. You can also access the PdAudioUnit `isBuffering` status property:

~~~
NSLog(@"is audio unit buffering? %@", (self.audioController.audioUnit.isBuffering ? "@yes" : @"no"));
~~~

So far this is working in my testing using many common sample rates for PdAudioUnit, ie. 8k, 16k, 24k, 22.05k, 11025, 44.1k, 48k, 88.2k, 96k. Also with Bluetooth which ends up needing an 8k or 16k external sample rate when grabbing from the microphone.

I am curios to know if libpd is now working seamlessly with Airpods, it should.

virusys commented:
@danomatika After a bit of testing today with all of the devices I listed above (minus the iPad 7th gen which I no longer have) here's what I found:

- on iPhone 6s if you connect an audio device (I used the Steinberg UR22mkII) with the Camera Connection Kit, launch pdtest, and then unplug the CCK, you will get clicking audio when it switches back to the iPhone's speakers. If you launch pdtest first you can plug/unplug the CCK as many times as you want with no issues. This is the only device I have that I could reproduce this on. 

- Selecting an Apple TV from the route picker doesn't seem to do anything; the only way I could stream audio from pdtest to either Apple TV (I currently have two I can test on) is by turning on Screen Mirroring but I'm not sure if that's what you're looking for. 

- Bluetooth seems to work fine on all devices I tried, great! (Although I should probably get a better pair of BT headphones to test with).

danomatika commented:
>on iPhone 6s if you connect an audio device (I used the Steinberg UR22mkII) with the Camera Connection Kit, launch pdtest, and then unplug the CCK, you will get clicking audio when it switches back to the iPhone's speakers. If you launch pdtest first you can plug/unplug the CCK as many times as you want with no issues. This is the only device I have that I could reproduce this on.

The sample rate conversion check may not be working. Can you try with buffering forced on? Comment or set buffer = YES here: https://github.com/libpd/libpd/blob/update/objc/objc/PdAudioUnit.m#L507

virusys commented:
> The sample rate conversion check may not be working. Can you try with buffering forced on? Comment or set buffer = YES here: https://github.com/libpd/libpd/blob/update/objc/objc/PdAudioUnit.m#L507

Setting buffer = YES seemed to do the trick! Works fine now.

danomatika commented:
Ok cool. Can you try with the latest commit? I think there was just a small boolean OR issue which is hopefully fixed. If not, then the buffering should simply be done whenever the sample rates are not equal instead of divisible.

Also, what sample rates were printed when `AU_DEBUG_VERBOSE` is defined?

virusys commented:
Here is my debugger output (I had to use Xcode's wireless debugging feature in order to get output while attached to the Camera Connection Kit) with `AU_DEBUG_VERBOSE` defined. Still produces the same clicking problem if you launch `pdtest` while the CCK is connected and then unplug the CCK. 

[pdtest output.txt](https://github.com/libpd/libpd/files/4136176/pdtest.output.txt)

I could be overlooking something pretty important though— I just remembered that this iPhone 6s is still running one of the later betas of iOS 13; I simply haven't had the time to free up enough space on the device to install the official release. If you think this is a problem let me know and I'll get on that right away.

danomatika commented:
The output is interesting in that the session is giving you 44.1k on a 6S whereas I could only get 48k when asking for 44.1k on the 6S I tested on which was also running iOS 12. Maybe it's something they changed with iOS 13?

virusys commented:
I just assumed it had something to do with the fact that I was launching pdtest while already connected to my Steinberg box (running at 44.1k) with the CCK— but when it was initially working it did print out 48k in the output.

danomatika commented:
I did some more work on this today and found the following:

Airplay
* works fine if the category is set for output (Playback, Ambient, etc)
* with the PlayAndRecored category works, ie. I had microphone input -> pd -> AirPlay
* I can select AirPlay outside of the app or via the route picker
* latency is big (not a surprise)

USB sample rate issue
* the problem comes from trying to detect if the number of channels is different when connecting/disconnecting a device and *not* also checking the sample rate, in this case IO channels was always 2 but the sample rate changed and buffer is not being started

Bluetooth sample rate issues
* I noticed that if Bluetooth wants 8k, the buffer size gets *very large* and there is stuttering, so buffering needs to be enabled regardless of sample rate conversion

After fiddling with sample rate & buffer detection, I think I'm at the point to where it's become too complicated to handle automatically using the buffer. The notification handling works, but there are a few edge cases where the audio unit can be created twice in a row (slow) or the session category loses it's options, etc. Now I think it's far simpler to always buffer and only reallocate when the number of input or output channels changes.

I did some basic performance tests and there isn't really a major difference between buffering or not buffering. It simply adds some memcpy steps in between input and output, so at this point I will probably simplify things by always having it on and remove the sample rate detection mechanisms. I may also refactor this buffering into an extra C layer that the other wrappers can use, such as Android.

cerupcat commented:
Agreed. I think always having the buffer will be helpful and may also help
in debugging (since fewer code paths).

Moving it to a C layer would also help so it could be factored into the
Android codebase and/or ofxPd since they'll all need this.
ᐧ

On Fri, Jan 31, 2020 at 8:51 AM Dan Wilcox <notifications@github.com> wrote:

> I did some more work on this today and found the following:
>
> Airplay
>
>    - works fine if the category is set for output (Playback, Ambient, etc)
>    - with the PlayAndRecored category works, ie. I had microphone input
>    -> pd -> AirPlay
>    - I can select AirPlay outside of the app or via the route picker
>    - latency is big (not a surprise)
>
> USB sample rate issue
>
>    - the problem comes from trying to detect if the number of channels is
>    different when connecting/disconnecting a device and *not* also
>    checking the sample rate, in this case IO channels was always 2 but the
>    sample rate changed and buffer is not being started
>
> Bluetooth sample rate issues
>
>    - I noticed that if Bluetooth wants 8k, the buffer size gets *very
>    large* and there is stuttering, so buffering needs to be enabled
>    regardless of sample rate conversion
>
> After fiddling with sample rate & buffer detection, I think I'm at the
> point to where it's become too complicated to handle automatically using
> the buffer. The notification handling works, but there are a few edge cases
> where the audio unit can be created twice in a row (slow) or the session
> category loses it's options, etc. Now I think it's far simpler to always
> buffer and only reallocate when the number of input or output channels
> changes.
>
> I did some basic performance tests and there isn't really a major
> difference between buffering or not buffering. It simply adds some memcpy
> steps in between input and output, so at this point I will probably
> simplify things by always having it on and remove the sample rate detection
> mechanisms. I may also refactor this buffering into an extra C layer that
> the other wrappers can use, such as Android.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/libpd/libpd/pull/283?email_source=notifications&email_token=AABR25366ZWWVKZIIJBGBQTRARJKRA5CNFSM4KHYIVW2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKPIUKA#issuecomment-580815400>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABR25ZTQFWAZNUKYUUUAYLRARJKRANCNFSM4KHYIVWQ>
> .
>

monkeyswarm commented:
Hi Dan, tested on
iPhone 6s - iOS 13.3
iPad mini 2 - iOS 12.4.4

Everything seems good. Tried plugging/unplugging headphones, changing sampling rates and buffer sizes, and connecting to bluetooth output. 

I used it on MobMuPlat; using PdAudioUnit as the superclass of MobMuPlatPdAudioUnit yielded some inscrutable crashes (when instantiating the latter), but using PdAudioUnit directly works fine. (Given that MobMuPlatPdAudioUnit was just a wrapper around an old version of Audiobus, I will get rid of it and use PdAudioUnit directly; I'll reconsider adding the current version of Audiobus in the future if people complain).

I wasn't able to test Bluetooth input, as MMP has an old implementation of categories that doesn't support it; I also wasn't able to test a USB audio device as I can't find my lightning-USB dongle.

Thanks for all your work on this.

monkeyswarm commented:
(I'll keep looking for older iOS version devices, and my lightning-USB dongle, and update if I find either)

danomatika commented:
Latest commits:

* simplify things by enabling audio unit buffering by default with an option to override while configuring and an audio controller override as well
*fixed slight Bluetooth buffer stutter by padding the max ring buffer sizes
* media services reset notification is now handled by caching settings and re-applying
* added empty silence secondary audio notification handler for easy subclassing

danomatika commented:
I decided to remove the optional automatic input and output channel handling, ie. pass -1 as number of channels will always run pd with whatever the device has. This was becoming problematic due to timing issues with notification handling so I think its better to do this explicitly outside of the audio controller. It also simplifies a few things and makes the route change transition smoother.

danomatika commented:
The only last issue I have is BluetoothA2DP devices (ie. HQ stereo output) not grabbing pdtest audio when connected while the app is running.The AVRoutePicker shows the device and I can select it, but nothing happens after the progress indicator spins a bit. If you start the app while a device is receiving the session audio, then connecting & disconnecting works fine, either automatic or via AVRoutePicker.

In any case, I'm not sure if this is an issue with any of the libpd objc code or just down to my old device talking to cheap Bluetooth speakers.

cornedriesprong commented:
I've tested it with our app and it fixes the audio glitches we were having with Airpods (tested with iPhone XR iOS 13.3.1).

I'm wondering if it would be possible to get this working with the `abl_link~` object from the pd-for-ios repo?

danomatika commented:
>I'm wondering if it would be possible to get this working with the abl_link~ object from the pd-for-ios repo?

I'm not sure how this branch is _incompatible_. You can build the abl_link~ sources into your project and include the link framework (or pre-compiled lib, however they provide it). The object should then be available within your libpd instance.

danomatika commented:
@cornedriesprong https://github.com/libpd/libpd/wiki/Adding-Pure-Data-external-libraries-to-your-project

cornedriesprong commented:
sorry, I wasn't being very clear. the [pd-for-ios](https://github.com/libpd/pd-for-ios/) repo contains a PdLinkAudioUnit (subclass of PdAudioUnit) which breaks with this update. I got abl_link~ working by integrating Link into your updated PdAudioUnit. I suppose I can file a pull on request on the pd-for-ios repo once this is merged.

danomatika commented:
It would be good to see what changes are needed. I’d like to find an easy way to inherit PdAudioUnit without having to basically copy/paste, ie. provide some kind of callback interface to make AudioBus, etc integration simple.

enohp ym morf tnes
-----------
Dan Wilcox
danomatika.com
robotcowboy.com


> On Feb 19, 2020, at 5:24 PM, Corné <notifications@github.com> wrote:
> 
> sorry, I wasn't being very clear. the pd-for-ios repo contains a PdLinkAudioUnit (subclass of PdAudioUnit) which breaks with this update. I got abl_link~ working by integrating Link into your updated PdAudioUnit. I suppose I can file a pull on request on the pd-for-ios repo once this is merged.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

BorisMolch commented:
Merging this to our iOS app fixed :
-Airpods producing harsh distortion
-No audio after background, foreground
-Distortion after sample rate change 

But - it introduced whats seems like memory allocation issues.
The crash is almost random, on app start.  (reproducible 5% of the time)

danomatika commented:
Any backtraces?

BorisMolch commented:
logs of some crashes we suspect to be related 
[;ibpd_related.zip](https://github.com/libpd/libpd/files/4358337/ibpd_related.zip)

danomatika commented:
The following could be a timing issue between threads which you are seeing on your newer, faster device (XS Max) and I don't on my older, slower device (5S).

~~~
#10. Crashed: Thread
EXC_BAD_ACCESS 0x00000000df3f0048
0  AudioToolboxCore               0x19e2fbb50 AudioUnitInitialize + 20
1  Jambl Prod                     0x101a32ab0 -[PdAudioUnit initAudioUnitWithSampleRate:inputChannels:outputChannels:] + 1244 (PdAudioUnit.m:481)
2  Jambl Prod                     0x101a31f20 -[PdAudioUnit configureWithSampleRate:inputChannels:outputChannels:bufferingEnabled:] + 336 (PdAudioUnit.m:140)
3  Jambl Prod                     0x101a31d48 -[PdAudioUnit handleRouteChange:] + 92 (PdAudioUnit.m:97)
4  CoreFoundation                 0x19118a288 __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 20
5  CoreFoundation                 0x19118a2d0 ___CFXRegistrationPost1_block_invoke + 64
6  CoreFoundation                 0x191189630 _CFXRegistrationPost1 + 368
7  CoreFoundation                 0x1911892e8 ___CFXNotificationPost_block_invoke + 104
8  CoreFoundation                 0x19110559c -[_CFXNotificationRegistrar find:object:observer:enumerator:] + 1416
9  CoreFoundation                 0x191188c38 _CFXNotificationPost + 1244
10 Foundation                     0x1914e50c8 -[NSNotificationCenter postNotificationName:object:userInfo:] + 60
11 AVFAudio                       0x19ddb2ec4 avfaudio::AVAudioSessionPropertyListener(void*, unsigned int, unsigned int, void const*) + 2656
12 AudioToolbox                   0x19c2897bc AudioSessionPropertyListeners::CallPropertyListeners(unsigned int, unsigned int, void const*) + 596
13 AudioToolbox                   0x19c2e2c3c HandleCFPropertyListChange(unsigned int, unsigned int, unsigned long, unsigned char*, unsigned int) + 716
14 AudioToolbox                   0x19c2e87b8 HandleAudioSessionCFTypePropertyChangedMessage(unsigned int, unsigned int, void*, unsigned int) + 328
15 AudioToolbox                   0x19c2e7eb4 ProcessDeferredMessage(unsigned int, __CFData const*, unsigned int, unsigned int) + 1620
...
~~~

However, I am unsure of which commit you are using from the branch as there is no `PdAudioUnit handleRouteChange:` method. If you are using an earlier commit, the problem is likely fixed already.

danomatika commented:
@BorisMolch This one could be an issue with a pd networking object. Is the patch that's being loaded use either [netsend] or [netreceive]?

~~~
#25. Crashed: Thread
EXC_BAD_ACCESS 0x000000006e6c2594
0  libobjc.A.dylib           0x1af5b60a0 objc_msgSend + 32
1  libnetwork.dylib          0x1b1d007b0 nw_protocol_options_access_handle + 96
2  libboringssl.dylib        0x1b46f7a04 boringssl_session_apply_protocol_options + 164
3  libboringssl.dylib        0x1b46fe8f0 nw_protocol_boringssl_connected + 1104
4  libusrtcp.dylib           0x1b4b33d54 nw_protocol_tcp_wake_connected + 396
5  libusrtcp.dylib           0x1b4b5b678 tcp_input_available + 764
6  libusrtcp.dylib           0x1b4b231e4 nw_protocol_tcp_input_available + 220
7  libnetwork.dylib          0x1b1c939fc nw_channel_add_input_frames + 4328
8  libnetwork.dylib          0x1b1c92898 nw_channel_update_input_source + 168
9  libnetwork.dylib          0x1b1c920c4 __nw_channel_create_block_invoke.22 + 52
10 libdispatch.dylib         0x1af55f184 _dispatch_client_callout + 16
11 libdispatch.dylib         0x1af507a3c _dispatch_continuation_pop$VARIANT$mp + 412
12 libdispatch.dylib         0x1af517fe4 _dispatch_source_invoke$VARIANT$mp + 1288
13 libdispatch.dylib         0x1af50cf7c _dispatch_workloop_invoke$VARIANT$mp + 1796
14 libdispatch.dylib         0x1af515314 _dispatch_workloop_worker_thread + 588
15 libsystem_pthread.dylib   0x1af5aef88 _pthread_wqthread + 276

#0. Thread
0  libsystem_malloc.dylib   0x1af589f58 default_zone_realloc + 76
1  libsystem_malloc.dylib   0x1af5882e0 realloc + 196
2  Jambl Prod               0x105a3cb10 resizebytes + 40 (m_memory.c:55)
3  Jambl Prod               0x105a37950 binbuf_restore + 84 (m_binbuf.c:362)
4  Jambl Prod               0x105a244c4 canvas_obj + 64 (g_text.c:202)
5  Jambl Prod               0x105a3bb3c pd_typedmess + 160 (m_class.c)
6  Jambl Prod               0x105a3831c binbuf_eval + 968 (m_binbuf.c:803)
7  Jambl Prod               0x105a3a9a8 binbuf_evalfile + 348 (m_binbuf.c:1507)
8  Jambl Prod               0x105a4785c do_create_abstraction + 328 (s_loader.c:459)
9  Jambl Prod               0x105a3bbfc pd_typedmess + 352 (m_class.c)
10 Jambl Prod               0x105a3b0d4 new_anything + 160 (m_class.c:904)
11 Jambl Prod               0x105a3bb3c pd_typedmess + 160 (m_class.c)
12 Jambl Prod               0x105a3831c binbuf_eval + 968 (m_binbuf.c:803)
13 Jambl Prod               0x105a24690 canvas_objtext + 104 (g_text.c:106)
14 Jambl Prod               0x105a3bb3c pd_typedmess + 160 (m_class.c)
15 Jambl Prod               0x105a3831c binbuf_eval + 968 (m_binbuf.c:803)
16 Jambl Prod               0x105a3a9a8 binbuf_evalfile + 348 (m_binbuf.c:1507)
17 Jambl Prod               0x105a4785c do_create_abstraction + 328 (s_loader.c:459)
18 Jambl Prod               0x105a3bbfc pd_typedmess + 352 (m_class.c)
19 Jambl Prod               0x105a3b0d4 new_anything + 160 (m_class.c:904)
20 Jambl Prod               0x105a3bb3c pd_typedmess + 160 (m_class.c)
21 Jambl Prod               0x105a3831c binbuf_eval + 968 (m_binbuf.c:803)
22 Jambl Prod               0x105a24690 canvas_objtext + 104 (g_text.c:106)
23 Jambl Prod               0x105a3bb3c pd_typedmess + 160 (m_class.c)
24 Jambl Prod               0x105a3831c binbuf_eval + 968 (m_binbuf.c:803)
25 Jambl Prod               0x105a3a9a8 binbuf_evalfile + 348 (m_binbuf.c:1507)
26 Jambl Prod               0x105a0444c glob_evalfile + 88 (g_canvas.c:2062)
27 Jambl Prod               0x105a6f70c libpd_openfile + 64 (z_libpd.c:129)
28 Jambl Prod               0x105a70f88 +[PdBase openFile:path:] + 232 (PdBase.m:269)
29 Jambl Prod               0x104ee6f44 specialized JBAudioController.loadPatch() + 196 (JBAudioController.swift)
30 Jambl Prod               0x104ee6a00 JBAudioController.init() + 364 (JBAudioController.swift:22)
31 Jambl Prod               0x104ee6b5c @objc JBAudioController.init() + 20 (<compiler-generated>)
32 Jambl Prod               0x104ff3250 audioController.set + 0 [inlined] (<compiler-generated>)
33 Jambl Prod               0x104ff3250 AppDelegate.application(_:didFinishLaunchingWithOptions:) + 2032 (AppDelegate.swift:125)
34 Jambl Prod               0x104ff3f7c @objc AppDelegate.application(_:didFinishLaunchingWithOptions:) + 156 (<compiler-generated>)
...
~~~

This could also be related to some issues on patch load which have been fixed with newer versions of Pd. I can't find the relevant issues right now, but if this is a crash report using an older version of libpd, it will probably fixed with an upgrade at some point.

elliott-io commented:
How can I build this for iOS?
